lib/ansible/modules/cloud/vmware/vmware_host_tcpip_stacks.py

```
@@ -0,0 +1,473 @@
+                            (host.name, to_native(not_supported.msg))
+                        )
+                    except vim.fault.HostConfigFault as config_fault:
+                        self.module.fail_json(
+                            msg="Failed to configure TCP/IP stacks for host '%s' due to : %s" %
+                            (host.name, to_native(config_fault.msg))
+                        )
+            else:
+                results['result'][host.name]['changed'] = False
+                message = 'All settings are already configured'
+            results['result'][host.name]['msg'] = message
+
+            host_change_list.append(changed)
+
+        if any(host_change_list):
+            results['changed'] = True
+        self.module.exit_json(**results)
+
+    def create_netstack_spec(self, instance, key, vmkernel_gateway, num_connections, congestion_algorithm):
+        """Create TCP/IP stack spec"""
+        changed = False
+        vmkernel_gateway_previous = num_connections_previous = congestion_algorithm_previous = None
+        netstack_spec = vim.host.NetworkConfig.NetStackSpec()
+        netstack_spec.operation = 'edit'
+        netstack_spec.netStackInstance = vim.host.NetStackInstance()
+        netstack_spec.netStackInstance.key = key
+        # The display name is only customizable with custom stacks
+        netstack_spec.netStackInstance.name = key
+        # DNS configuration is only supported for the default TCP/IP stack or custom stacks
+        # netstack_spec.netStackInstance.dnsConfig = vim.host.DnsConfig()
+        ip_address = None
+        if vmkernel_gateway:
+            try:
+                # string need to be unicode; otherwise, the command will fail
+                ip_address = ipaddress.ip_address(text_type(vmkernel_gateway))
+            except ipaddress.AddressValueError:
+                self.module.fail_json(msg="%s is not a valid IP address!" % vmkernel_gateway)
```

Input validation should be done before doing API calls to fetch information (you already fetch the hosts and the network stacks before we hit this). It's better to fail early and quick...

---

Why did it even get to the module at all?  If there was an ip_addr (or ip_addr_v4, or ip_addr_v6) field type in the ansible core validator for module inputs, ansible would have rejected it before loading the module.

---

I need to check if it's an IPv4 or IPv6 and use the proper API parameter automatically. I don't want to use two different options for IPv4 or IPv6. It's just an except.

---

I was suggesting the ansible core could add an ipaddr argument type and validate it early enough that the concern from @pdellaert would be addressed both here and elsewhere in an easier way for everyone.  It doesn't have to be done as part of this PR.

---

`Provide functions for IP address or port range validating` https://github.com/ansible/ansible/pull/17731 may partly help with this.